File: advanced/account-compression/counter-compressed/programs/counter-compressed/src/counter.rs
use anchor_lang::prelude::*;
use anchor_lang::solana_program::keccak;

#[derive(Copy, Debug)]
#[account]
pub struct Counter {
    /// This identifies which tree it belongs to
    pub tree: Pubkey,
    /// This keeps track of which counter it is in the global state tree
    pub id: u32,
    /// Our normal count incrementer
    pub count: u64,
}

pub const COUNTER_PREFIX: &str = "counter";
pub const COUNTER_SIZE: usize = 8 + 32 + 4 + 8;

impl Counter {
    pub fn new(tree: &Pubkey, id: u32) -> Counter {
        Counter {
            tree: *tree,
            id,
            count: 0,
        }
    }
}

pub fn hash_counter(counter: Counter) -> Result<[u8; 32]> {
    Ok(keccak::hashv(&[&counter.try_to_vec()?]).to_bytes())
}

File: basics/counter/poseidon/ts-programs/src/counter.ts
import { Account, Pubkey, type Result, Signer, u64 } from '@solanaturbine/poseidon';

export default class Counter {
  static PROGRAM_ID = new Pubkey('GkV2QQhYjUdtodnREvNATozYWoEsTQX8ezUZs8Ncy3gm');

  initializeCounter(payer: Signer, counter: CounterState): Result {
    counter.derive(['']).init();
    counter.count = new u64(0);
  }

  incrementCounter(counter: CounterState): Result {
    counter.derive(['']);
    counter.count = counter.count.add(1);
  }
}

export interface CounterState extends Account {
  count: u64;
}
